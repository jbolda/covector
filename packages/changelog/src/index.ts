import { type Operation, all } from "effection";
import { readAllChangelogs } from "./get";
import { writeAllChangelogs } from "./write";
import unified from "unified";
import parse from "remark-parse";
import stringify from "remark-stringify";

import type {
  File,
  ConfigFile,
  PkgCommandResponse,
  AssembledChanges,
  Meta,
  ChangeContext,
} from "@covector/types";

export { pullLastChangelog } from "./get";

export function* fillChangelogs({
  applied,
  assembledChanges,
  config,
  cwd,
  pkgCommandsRan,
  create = true,
  createContext,
}: {
  applied: { name: string; version: string }[];
  assembledChanges: AssembledChanges;
  config: ConfigFile;
  cwd: string;
  pkgCommandsRan?: { [k: string]: PkgCommandResponse };
  create?: boolean;
  createContext?: ChangeContext;
}): Operation<{ [k: string]: PkgCommandResponse } | undefined> {
  const changelogs = yield readAllChangelogs({
    applied: applied.reduce(
      (
        final: { name: string; version: string; changelog?: File }[],
        current
      ) =>
        !config.packages[current.name].path ? final : final.concat([current]),
      []
    ),
    packages: config.packages,
    cwd,
  });

  const writtenChanges: ChangedLog[] = yield applyChanges({
    changelogs,
    assembledChanges,
    config,
    applied,
    createContext,
  });

  if (create) {
    yield writeAllChangelogs({ writtenChanges, cwd });
  }

  if (!pkgCommandsRan) {
    return;
  } else {
    pkgCommandsRan = Object.keys(pkgCommandsRan).reduce(
      (pkgs: { [k: string]: PkgCommandResponse }, pkg) => {
        writtenChanges.forEach((change) => {
          if (change.pkg === pkg) {
            pkgs[pkg].command = change.addition;
          }
        });
        return pkgs;
      },
      pkgCommandsRan
    );

    return pkgCommandsRan;
  }
}

export const pipeChangelogToCommands = async ({
  changelogs,
  pkgCommandsRan,
}: {
  changelogs: { [k: string]: { pkg: string; changelog: string } };
  pkgCommandsRan: { [k: string]: PkgCommandResponse };
}) =>
  Object.keys(pkgCommandsRan).reduce(
    (pkgs: { [k: string]: PkgCommandResponse }, pkg) => {
      Object.keys(changelogs).forEach((pkg) => {
        if (pkgs[pkg]) {
          pkgs[pkg].command = changelogs[pkg].changelog;
        }
      });
      return pkgs;
    },
    pkgCommandsRan
  );

const getVersionFromApplied = (
  name: string,
  applied: { name: string; version: string }[]
) => applied.find((pkg) => pkg.name === name)?.version;

/**
 *  Renders a change file in the format:
 *
 * - [`<commit-hash>`](<commit-url>)([#<pr-number>](<pr-url>)) change file summary
 *
 */
const renderRelease = (
  context: Record<string, Record<string, string>>,
  gitSiteUrl: string,
  release: {
    summary: string;
    meta?: Meta;
  },
  indentation: number = 4
) => {
  // indent the summary so it fits under the new bullet point we added
  const summary = release.summary.replace(
    /\n/g,
    `\n${" ".repeat(indentation)}`
  );

  if (!release.meta || (!!release.meta && !release.meta.commits)) {
    return `\n- ${summary}`;
  } else {
    const commit = release.meta.commits![0];
    const commitLink = `[\`${commit.hashShort}\`](${gitSiteUrl}commit/${commit.hashLong})`;
    const matches = commit.commitSubject.match(/(#[0-9]+)/g) ?? [];
    const len = matches.length;
    const pr = len === 0 ? null : len === 1 ? matches[0] : matches[len - 1];
    const hashLookup = `sha_${commit.hashLong}`;
    const prLink = pr ? `[${pr}](${gitSiteUrl}pull/${pr.slice(1)})` : "";
    const personLink = (person: string) =>
      `[@${person}](${gitSiteUrl}../../${person})`;
    const recognizeContributions = context?.[hashLookup]?.author
      ? ` by ${personLink(context[hashLookup].author)}`
      : ``;

    return `\n- ${commitLink}${
      pr ? ` (${prLink}${recognizeContributions})` : ""
    } ${summary}`;
  }
};

type Change = {
  changes: { name: string; version: string };
  changelog?: File;
};
type ChangedLog = { pkg: string; change: Change; addition: string };

function* defaultCreateContext(): Operation<
  Operation<{
    context: Record<string, string>;
    changeContext: Record<string, string>;
  }>
> {
  const context = {};
  return function* defineContexts(): Operation<{
    context: Record<string, string>;
    changeContext: Record<string, string>;
  }> {
    const changeContext = {};
    return { context, changeContext };
  };
}

function* applyChanges({
  changelogs,
  assembledChanges,
  config,
  applied,
  createContext = defaultCreateContext,
}: {
  changelogs: Change[];
  assembledChanges: AssembledChanges;
  config: ConfigFile;
  applied: { name: string; version: string }[];
  createContext?: ChangeContext;
}): Operation<ChangedLog[]> {
  const gitSiteUrl = !config.gitSiteUrl
    ? "/"
    : config.gitSiteUrl.replace(/\/$/, "") + "/";

  const processor: any = unified().use(parse).use(stringify, {
    bullet: "-",
    listItemIndent: "one",
  });

  const commits = [
    ...new Set(
      Object.values(assembledChanges.releases).flatMap((release) =>
        release.changes
          .map((change) => change.meta?.commits?.[0].hashLong)
          .filter(Boolean)
      )
    ),
  ];
  // @ts-expect-error expression not callable, but it is, we don't have a reasonable way to type narrow
  //  through the exported types from effection however (expects that it could possibly be OperationPromise)
  const createChangeContext = yield createContext({ commits });

  return yield all(
    changelogs.map(function* (change) {
      let additionChunks = [];
      if (change.changelog) {
        let changelog = processor.parse(change.changelog.content);
        if (!assembledChanges.releases[change.changes.name]) {
          additionChunks.push(
            `## [${change.changes.version}]\nBumped due to dependency.`
          );
        } else {
          additionChunks.push(`## [${change.changes.version}]`);

          const changeTags: { [k: string]: string } = config.changeTags ?? {};
          // ensures there is a `deps` tag if one wasn't defined
          if (!("deps" in changeTags)) {
            changeTags.deps = "Dependencies";
          }

          /**
           * Untagged changes are changes that don't have a tag associated with and `config.defaultChangeTag` is not set.
           * These will be rendered without a tag (section or category) at the beginning of a release changelog.
           */
          const untaggedChanges: {
            summary: string;
            meta?: Meta | undefined;
            tag?: string | undefined;
          }[] = [];
          const groupedChangesByTag: {
            [k: string]: {
              summary: string;
              meta?: Meta | undefined;
              tag?: string | undefined;
            }[];
          } = {};

          // **IMPORTANT**: prefill the `groupedChangesByTag` with tags from config
          // in the same order they were defined
          Object.keys(changeTags).forEach((k) => (groupedChangesByTag[k] = []));

          // fill `deps` tag
          const dependencies = Object.keys(
            assembledChanges.releases[change.changes.name].changes
              .filter((c) => c.meta?.dependencies)
              // reduce to an object of keys to avoid duplication of deps
              .reduce(
                (acc, c) => {
                  c.meta!.dependencies.forEach((dep) => (acc[dep] = 1));
                  return acc;
                },
                {} as { [k: string]: any }
              )
          ).map((dep) => {
            const appliedVersion = getVersionFromApplied(dep, applied);
            return {
              summary: appliedVersion
                ? `Upgraded to \`${dep}@${appliedVersion}\``
                : `Upgraded to latest \`${dep}\``,
            };
          });
          groupedChangesByTag.deps.push(...dependencies);

          assembledChanges.releases[change.changes.name].changes
            .filter((c) => !c.meta?.dependencies)
            .forEach((c) => {
              // fallback to `defaultChangeTag` if it is set, otherwise mark this change as untagged
              const key = c.tag ?? config.defaultChangeTag;
              if (key) {
                if (!groupedChangesByTag[key]) groupedChangesByTag[key] = [];
                groupedChangesByTag[key].push(c);
              } else {
                untaggedChanges.push(c);
              }
            });

          const {
            context,
          }: { context: Record<string, Record<string, string>> } =
            yield createChangeContext();

          // render untagged changes freely at the top
          for (const release of untaggedChanges) {
            additionChunks.push(renderRelease(context, gitSiteUrl, release));
          }

          // render tagged changes as:
          //
          // ### <Long tag1 name>
          //
          // - `1534ae12`(#12) Change1 summary
          // - `1534ae12`(#12) Change2 summary
          //
          // ### <Long tag2 name>
          //
          // - `1534ae12`(#12) Change1 summary
          // - `1534ae12`(#12) Change2 summary
          //
          // ...etc
          for (const [tag, change] of Object.entries(groupedChangesByTag)) {
            if (change.length !== 0) {
              // if the specified tag is not defined in config,
              // fallback to the short tag specified in the change file
              const longTag = changeTags[tag] ?? tag;
              additionChunks.push(`\n### ${longTag}\n`);
              for (const release of change) {
                additionChunks.push(
                  renderRelease(context, gitSiteUrl, release)
                );
              }
            }
          }
        }

        const addition = additionChunks.join("");
        const parsedAddition = processor.parse(addition);
        const changelogFirstElement = changelog.children.shift();
        const changelogRemainingElements = changelog.children;
        changelog.children = [].concat(
          changelogFirstElement,
          parsedAddition.children,
          changelogRemainingElements
        );
        change.changelog.content = processor.stringify(changelog);
      }
      return {
        pkg: change.changes.name,
        change,
        addition: additionChunks.join(""),
      };
    })
  );
}
