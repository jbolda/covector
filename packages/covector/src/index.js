const { covector } = require("./run");
module.exports.covector = covector;

const { cli } = require("./cli");
module.exports.cli = cli;
