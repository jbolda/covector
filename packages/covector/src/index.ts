import { covector } from "./run";
import { cli } from "./cli";

export { covector, cli };
