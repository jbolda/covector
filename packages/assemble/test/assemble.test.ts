import { assemble } from "../src";
import { captureError, describe, it } from "../../../helpers/test-scope.ts";
import { expect } from "vitest";
import pino from "pino";
import * as pinoTest from "pino-test";

const filePart = (filename: string) => ({
  filename,
  path: `.changes/${filename}`,
  extname: "",
  content: "",
});

const config = {
  pkgManagers: {
    javascript: {
      version: "lerna version ${ release.type } --no-git-tag-version --no-push",
      publish: "npm publish",
    },
  },
  packages: {
    assemble1: {
      path: "./packages/assemble1",
      manager: "javascript",
      dependencies: ["all"],
    },
    assemble2: {
      path: "./packages/assemble2",
      version: "lerna version ${ release.type }",
      dependencies: ["all"],
    },
    "@namespaced/assemble1": {
      path: "./packages/namespaced-assemble2",
      manager: "cargo",
      version: "cargo version ${ release.type }",
      publish: "cargo publish",
      dependencies: ["assemble1", "all"],
    },
    "@namespaced/assemble2": {
      path: "./packages/namespaced-assemble2",
      manager: "cargo",
      version: "cargo version ${ release.type }",
      publish: "cargo publish",
      dependencies: ["assemble2", "all"],
    },
    all: {
      version: true,
    },
  },
};

const configSpecial = {
  ...config,
  additionalBumpTypes: ["housekeeping", "workflows"],
};

const testTextOne = {
  ...filePart("testTextOne"),
  content: `
---
"assemble1": patch
"assemble2": patch
---
    
This is a test.
`,
};
const testTextTwo = {
  ...filePart("testTextTwo"),
  content: `
---
"assemble1": minor
"assemble2": patch
---
    
This is a test.
`,
};
const testTextThree = {
  ...filePart("testTextThree"),
  content: `
---
"assemble1": patch
"assemble2": major
---
    
This is a test.
`,
};
const testTextFour = {
  ...filePart("testTextFour"),
  content: `
---
"assemble1": patch
"@namespaced/assemble2": patch
---
    
This is a test. We might link out to a [website](https://www.jacobbolda.com).
`,
};
const testTextFive = {
  ...filePart("testTextFive"),
  content: `
---
"all": minor
---
  
This is a test.
`,
};
const testTextSpecialOne = {
  ...filePart("testTextSpecialOne"),
  content: `
---
"assemble1": housekeeping
"assemble2": workflows
---
  
This is a test.
`,
};
const testTextSpecialTwo = {
  ...filePart("testTextSpecialTwo"),
  content: `
---
"assemble1": patch
"assemble2": workflows
"@namespaced/assemble2": explosions
---
  
This is a test.
`,
};

describe("assemble", () => {
  describe("assemble changes", () => {
    it("runs", function* () {
      const stream = pinoTest.sink();
      const logger = pino(stream);

      const assembled = yield assemble({
        logger,
        files: [testTextOne, testTextTwo, testTextThree, testTextFour],
      });
      expect(assembled).toMatchSnapshot();
    });

    it("assembles deps", function* () {
      const stream = pinoTest.sink();
      const logger = pino(stream);

      const assembled = yield assemble({ logger, files: [testTextFive] });
      expect(assembled).toMatchSnapshot();
    });
  });

  describe("assemble changes in preMode", () => {
    it("with no existing changes", function* () {
      const stream = pinoTest.sink();
      const logger = pino(stream);

      const assembled = yield assemble({
        logger,
        files: [testTextOne, testTextTwo, testTextThree, testTextFour],
        preMode: { on: true, prevFiles: [] },
      });
      expect(assembled.releases).toMatchObject({
        "@namespaced/assemble2": { type: "prepatch" },
      });
      expect(assembled.releases).toMatchObject({
        assemble1: { type: "preminor" },
      });
      expect(assembled.releases).toMatchObject({
        assemble2: { type: "premajor" },
      });
      expect(assembled).toMatchSnapshot();
    });

    it("with existing changes that upgrade", function* () {
      const stream = pinoTest.sink();
      const logger = pino(stream);

      const assembled = yield assemble({
        logger,
        files: [testTextOne, testTextTwo, testTextThree, testTextFour],
        preMode: { on: true, prevFiles: [testTextOne.path] },
      });
      expect(assembled.changes).not.toEqual(
        expect.arrayContaining([
          {
            releases: { assemble1: "patch", assemble2: "patch" },
            summary: "This is a test.",
          },
        ])
      );
      expect(assembled.releases).toMatchObject({
        "@namespaced/assemble2": { type: "prepatch" },
      });
      expect(assembled.releases).toMatchObject({
        assemble1: { type: "preminor" },
      });
      expect(assembled.releases).toMatchObject({
        assemble2: { type: "premajor" },
      });
      expect(assembled).toMatchSnapshot();
    });

    it("with existing changes with the same bump", function* () {
      const stream = pinoTest.sink();
      const logger = pino(stream);

      const assembled = yield assemble({
        logger,
        files: [testTextOne, testTextTwo, testTextFour],
        preMode: { on: true, prevFiles: [testTextOne.path] },
      });
      expect(assembled.changes).not.toEqual(
        expect.arrayContaining([
          {
            releases: { assemble1: "patch", assemble2: "patch" },
            summary: "This is a test.",
          },
        ])
      );
      expect(assembled.releases).toMatchObject({
        "@namespaced/assemble2": { type: "prepatch" },
      });
      expect(assembled.releases).toMatchObject({
        assemble1: { type: "preminor" },
      });
      expect(assembled.releases).toMatchObject({
        assemble2: { type: "prerelease" },
      });
      expect(assembled).toMatchSnapshot();
    });

    it("with existing changes and a first bump", function* () {
      const stream = pinoTest.sink();
      const logger = pino(stream);

      const assembled = yield assemble({
        logger,
        files: [testTextOne, testTextTwo],
        preMode: { on: true, prevFiles: [testTextOne.path] },
      });
      expect(assembled.changes).not.toEqual(
        expect.arrayContaining([
          {
            releases: { assemble1: "patch", assemble2: "patch" },
            summary: "This is a test.",
          },
        ])
      );
      expect(assembled.releases).toMatchObject({
        assemble1: { type: "preminor" },
      });
      expect(assembled.releases).toMatchObject({
        assemble2: { type: "prerelease" },
      });
      expect(assembled).toMatchSnapshot();
    });

    it("with existing changes and a lower bump", function* () {
      const stream = pinoTest.sink();
      const logger = pino(stream);

      const assembled = yield assemble({
        logger,
        files: [testTextOne, testTextTwo],
        preMode: { on: true, prevFiles: [testTextTwo.path] },
      });
      expect(assembled.changes).not.toEqual(
        expect.arrayContaining([
          {
            releases: { assemble1: "minor", assemble2: "patch" },
            summary: "This is a test.",
          },
        ])
      );
      expect(assembled.releases).toMatchObject({
        assemble1: { type: "prerelease" },
      });
      expect(assembled.releases).toMatchObject({
        assemble2: { type: "prerelease" },
      });
      expect(assembled).toMatchSnapshot();
    });
  });

  describe("errors on bad change files", () => {
    const emptyChangefile = {
      path: ".changes/empty-file.md",
      extname: "",
      filename: "empty-file.md",
      content: `---
---
  
This doesn't bump much.
`,
    };

    it("throws on no changes", function* () {
      const stream = pinoTest.sink();
      const logger = pino(stream);

      expect.assertions(1);
      const e = yield captureError(
        assemble({
          logger,
          files: [emptyChangefile],
        })
      );
      expect(e.message).toMatch(
        ".changes/empty-file.md didn't have any packages bumped. Please add a package bump."
      );
    });
  });

  describe("special bump types", () => {
    it("valid additional bump types", function* () {
      const stream = pinoTest.sink();
      const logger = pino(stream);

      const assembled = yield assemble({
        logger,
        files: [
          testTextOne,
          testTextTwo,
          testTextThree,
          testTextFour,
          testTextSpecialOne,
        ],
        config: configSpecial,
      });
      expect(assembled).toMatchSnapshot();
    });

    it("invalid bump types", function* () {
      const stream = pinoTest.sink();
      const logger = pino(stream);

      expect.assertions(1);
      const e = yield captureError(
        assemble({
          logger,
          files: [
            testTextOne,
            testTextTwo,
            testTextThree,
            testTextFour,
            testTextSpecialOne,
          ],
          config,
        })
      );
      expect(e.message).toMatch(
        "housekeeping specified for assemble1 is invalid.\n" +
          "Try one of the following: major, minor, patch.\n"
      );
    });

    it("one each valid and invalid", function* () {
      const stream = pinoTest.sink();
      const logger = pino(stream);

      expect.assertions(1);
      const e = yield captureError(
        assemble({
          logger,
          files: [testTextSpecialTwo],
          config: configSpecial,
        })
      );
      expect(e.message).toMatch(
        "explosions specified for @namespaced/assemble2 is invalid.\n" +
          "Try one of the following: major, minor, patch, housekeeping, workflows.\n"
      );
    });

    it("handles an only noop", function* () {
      const stream = pinoTest.sink();
      const logger = pino(stream);

      const assembled = yield assemble({
        logger,
        files: [testTextSpecialOne],
        config: configSpecial,
      });
      expect(assembled).toMatchSnapshot();
    });
  });
});
