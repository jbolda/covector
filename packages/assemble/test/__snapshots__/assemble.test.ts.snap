// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`assemble > assemble changes > assembles deps 1`] = `
{
  "changes": [
    {
      "releases": {
        "all": "minor",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
  ],
  "releases": {
    "all": {
      "changes": [
        {
          "releases": {
            "all": "minor",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
      ],
      "type": "minor",
    },
  },
}
`;

exports[`assemble > assemble changes > runs 1`] = `
{
  "changes": [
    {
      "releases": {
        "assemble1": "patch",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
    {
      "releases": {
        "assemble1": "minor",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
    {
      "releases": {
        "assemble1": "patch",
        "assemble2": "major",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
    {
      "releases": {
        "@namespaced/assemble2": "patch",
        "assemble1": "patch",
      },
      "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
      "tag": undefined,
    },
  ],
  "releases": {
    "@namespaced/assemble2": {
      "changes": [
        {
          "releases": {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
          "tag": undefined,
        },
      ],
      "type": "patch",
    },
    "assemble1": {
      "changes": [
        {
          "releases": {
            "assemble1": "patch",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "assemble1": "patch",
            "assemble2": "major",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
          "tag": undefined,
        },
      ],
      "type": "minor",
    },
    "assemble2": {
      "changes": [
        {
          "releases": {
            "assemble1": "patch",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "assemble1": "patch",
            "assemble2": "major",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
      ],
      "type": "major",
    },
  },
}
`;

exports[`assemble > assemble changes in preMode > with existing changes and a first bump 1`] = `
{
  "changes": [
    {
      "releases": {
        "assemble1": "minor",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
  ],
  "releases": {
    "assemble1": {
      "changes": [
        {
          "releases": {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
      ],
      "type": "preminor",
    },
    "assemble2": {
      "changes": [
        {
          "releases": {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
      ],
      "type": "prerelease",
    },
  },
}
`;

exports[`assemble > assemble changes in preMode > with existing changes and a lower bump 1`] = `
{
  "changes": [
    {
      "releases": {
        "assemble1": "patch",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
  ],
  "releases": {
    "assemble1": {
      "changes": [
        {
          "releases": {
            "assemble1": "patch",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
      ],
      "type": "prerelease",
    },
    "assemble2": {
      "changes": [
        {
          "releases": {
            "assemble1": "patch",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
      ],
      "type": "prerelease",
    },
  },
}
`;

exports[`assemble > assemble changes in preMode > with existing changes that upgrade 1`] = `
{
  "changes": [
    {
      "releases": {
        "assemble1": "minor",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
    {
      "releases": {
        "assemble1": "patch",
        "assemble2": "major",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
    {
      "releases": {
        "@namespaced/assemble2": "patch",
        "assemble1": "patch",
      },
      "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
      "tag": undefined,
    },
  ],
  "releases": {
    "@namespaced/assemble2": {
      "changes": [
        {
          "releases": {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
          "tag": undefined,
        },
      ],
      "type": "prepatch",
    },
    "assemble1": {
      "changes": [
        {
          "releases": {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "assemble1": "patch",
            "assemble2": "major",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
          "tag": undefined,
        },
      ],
      "type": "preminor",
    },
    "assemble2": {
      "changes": [
        {
          "releases": {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "assemble1": "patch",
            "assemble2": "major",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
      ],
      "type": "premajor",
    },
  },
}
`;

exports[`assemble > assemble changes in preMode > with existing changes with the same bump 1`] = `
{
  "changes": [
    {
      "releases": {
        "assemble1": "minor",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
    {
      "releases": {
        "@namespaced/assemble2": "patch",
        "assemble1": "patch",
      },
      "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
      "tag": undefined,
    },
  ],
  "releases": {
    "@namespaced/assemble2": {
      "changes": [
        {
          "releases": {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
          "tag": undefined,
        },
      ],
      "type": "prepatch",
    },
    "assemble1": {
      "changes": [
        {
          "releases": {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
          "tag": undefined,
        },
      ],
      "type": "preminor",
    },
    "assemble2": {
      "changes": [
        {
          "releases": {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
      ],
      "type": "prerelease",
    },
  },
}
`;

exports[`assemble > assemble changes in preMode > with no existing changes 1`] = `
{
  "changes": [
    {
      "releases": {
        "assemble1": "patch",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
    {
      "releases": {
        "assemble1": "minor",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
    {
      "releases": {
        "assemble1": "patch",
        "assemble2": "major",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
    {
      "releases": {
        "@namespaced/assemble2": "patch",
        "assemble1": "patch",
      },
      "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
      "tag": undefined,
    },
  ],
  "releases": {
    "@namespaced/assemble2": {
      "changes": [
        {
          "releases": {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
          "tag": undefined,
        },
      ],
      "type": "prepatch",
    },
    "assemble1": {
      "changes": [
        {
          "releases": {
            "assemble1": "patch",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "assemble1": "patch",
            "assemble2": "major",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
          "tag": undefined,
        },
      ],
      "type": "preminor",
    },
    "assemble2": {
      "changes": [
        {
          "releases": {
            "assemble1": "patch",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "assemble1": "patch",
            "assemble2": "major",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
      ],
      "type": "premajor",
    },
  },
}
`;

exports[`assemble > special bump types > handles an only noop 1`] = `
{
  "changes": [
    {
      "releases": {
        "assemble1": "housekeeping",
        "assemble2": "workflows",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
  ],
  "releases": {
    "assemble1": {
      "changes": [
        {
          "releases": {
            "assemble1": "housekeeping",
            "assemble2": "workflows",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
      ],
      "type": "noop",
    },
    "assemble2": {
      "changes": [
        {
          "releases": {
            "assemble1": "housekeeping",
            "assemble2": "workflows",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
      ],
      "type": "noop",
    },
  },
}
`;

exports[`assemble > special bump types > valid additional bump types 1`] = `
{
  "changes": [
    {
      "releases": {
        "assemble1": "patch",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
    {
      "releases": {
        "assemble1": "minor",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
    {
      "releases": {
        "assemble1": "patch",
        "assemble2": "major",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
    {
      "releases": {
        "@namespaced/assemble2": "patch",
        "assemble1": "patch",
      },
      "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
      "tag": undefined,
    },
    {
      "releases": {
        "assemble1": "housekeeping",
        "assemble2": "workflows",
      },
      "summary": "This is a test.",
      "tag": undefined,
    },
  ],
  "releases": {
    "@namespaced/assemble2": {
      "changes": [
        {
          "releases": {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
          "tag": undefined,
        },
      ],
      "type": "patch",
    },
    "assemble1": {
      "changes": [
        {
          "releases": {
            "assemble1": "patch",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "assemble1": "patch",
            "assemble2": "major",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
          "tag": undefined,
        },
        {
          "releases": {
            "assemble1": "housekeeping",
            "assemble2": "workflows",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
      ],
      "type": "minor",
    },
    "assemble2": {
      "changes": [
        {
          "releases": {
            "assemble1": "patch",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "assemble1": "patch",
            "assemble2": "major",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
        {
          "releases": {
            "assemble1": "housekeeping",
            "assemble2": "workflows",
          },
          "summary": "This is a test.",
          "tag": undefined,
        },
      ],
      "type": "major",
    },
  },
}
`;
